advanced-programming/cgo/rocks/rocks.go

```diff
@@ -0,0 +1,83 @@
+package rocks
+
+/*
+#cgo CFLAGS: -I/opt/homebrew/Cellar/rocksdb/6.20.3/include
+#cgo LDFLAGS: -L/opt/homebrew/Cellar/rocksdb/6.20.3/lib -lrocksdb -lz -lbz2 -lpthread
+#include <stdlib.h>
+#include <strings.h>
+#include <rocksdb/c.h>
+*/
+import "C"
+import (
+	"fmt"
+	"unsafe"
+)
+
+type database struct {
+	db           *C.rocksdb_t
+	options      *C.rocksdb_options_t
+	writeOptions *C.rocksdb_writeoptions_t
+	readOptions  *C.rocksdb_readoptions_t
+	err          *C.char
+}
+
+func (db *database) Get(key string) (string, error) {
+	length := C.size_t(0)
+	k := C.CString(key)
+	v := C.rocksdb_get(db.db, db.readOptions, k, C.strlen(k), &length, &db.err)
+	defer func() {
+		C.free(unsafe.Pointer(k))
+		C.free(unsafe.Pointer(v))
+	}()
+
+	if db.err != nil {
```

Comment:
I'm a bit surprised that this check works - I would expect that you need to dereference the pointer and perform a string comparison here against the empty string since line 61 guarantees the pointer itself isn't nil.

What am I missing?

Reply:
Hmm, good question. After a quick test I think it is `nil` when it's defined though.

I just added this after line 61:
```go
fmt.Printf("%v\n", err)
```

And it does print `<nil>`. I'm also able to give bad input and get the error message.